File: AjaxVectorObjects/Common/RedoUndo/LayerAddedCommand.cs
// Copyright (c) 2018 Aurigma Inc. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
//

namespace Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo
{
    internal class LayerAddedCommand : Command
    {
        private int _layerIndex;
        private LayerData _layerData;

        public LayerAddedCommand(Layer layer, int index)
        {
            _layerIndex = index;
            _layerData = new LayerData(layer);
        }

        public LayerAddedCommand()
        {
            _layerIndex = -1;
            _layerData = new LayerData();
        }

        public int LayerIndex
        {
            get
            {
                return _layerIndex;
            }
            set
            {
                _layerIndex = value;
            }
        }

        public LayerData LayerData
        {
            get
            {
                return _layerData;
            }
            set
            {
                _layerData = value;
            }
        }

        public override void Execute(ICanvas canvas)
        {
            Layer l = new Layer();
            this.LayerData.ApplyState(l);
            canvas.Layers.Insert(this.LayerIndex, l);
        }

        public override void UnExecute(ICanvas canvas)
        {
            this.LayerData = new LayerData(canvas.Layers[this.LayerIndex]);
            canvas.Layers.RemoveAt(this.LayerIndex);
        }
    }
}

File: AjaxVectorObjects/ClientScripts/RedoUndo/LayerAddedCommand.js
// Copyright (c) 2018 Aurigma Inc. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
//
Type.registerNamespace("Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo");

Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.LayerAddedCommand = function (layer, index) {
    /// <summary>This class represents the layer added command.</summary>
    /// <constructor>
    /// 	<exclude />
    /// </constructor>
    this.LayerIndex = -1;
    this.LayerData = new Aurigma.GraphicsMill.AjaxControls.VectorObjects.LayerData();
    if (layer) {
        this.LayerData = new Aurigma.GraphicsMill.AjaxControls.VectorObjects.LayerData(layer);
    }
    if (index) {
        this.LayerIndex = index;
    }
}

Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.LayerAddedCommand.prototype = {
    execute: function (canvas) {
        /// <summary>Executes this command.</summary>
        /// <param name="canvas" type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.Canvas">The canvas to execute the command on.</param>
        var l = new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Layer(canvas);
        Aurigma.GraphicsMill.AjaxControls.VectorObjects.LayerData.applyState(this.LayerData, l);
        canvas.get_layers().insert(this.LayerIndex, l);

        Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.LayerAddedCommand.callBaseMethod(this, 'execute');
    },

    unExecute: function (canvas) {
        /// <summary>Rolls this command back.</summary>
        /// <param name="canvas" type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.Canvas">The canvas to roll the command back.</param>
        this.LayerData = new Aurigma.GraphicsMill.AjaxControls.VectorObjects.LayerData(canvas.get_layers().get_item(this.LayerIndex));
        canvas.get_layers().removeAt(this.LayerIndex);

        Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.LayerAddedCommand.callBaseMethod(this, 'unExecute');
    }
}

Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.LayerAddedCommand.registerClass("Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.LayerAddedCommand",
	Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.Command);